src/math_utils.rs

```diff
@@ -1,7 +1,29 @@
-use crate::bigint;
+use std::ops::Shr;
+
+use crate::{bigint, vm::errors::vm_errors::VirtualMachineError};
 use num_bigint::BigInt;
 use num_integer::Integer;
-use num_traits::{abs, FromPrimitive};
+use num_traits::{abs, FromPrimitive, Signed};
+
+///Returns the integer square root of the nonnegative integer n.
+///This is the floor of the exact square root of n.
+///Unlike math.sqrt(), this function doesn't have rounding error issues.
+pub fn isqrt(n: &BigInt) -> Result<BigInt, VirtualMachineError> {
+    //n.shr(1) = n.div_floor(2)
+    if !n.is_positive() {
+        return Err(VirtualMachineError::SqrtNegative(n.clone()));
+    }
+    let mut x = n.clone();
+    let mut y = (x.clone() + bigint!(1)).shr(1_i32);
+    while y < x.clone() {
+        x = y;
+        y = (x.clone() + n.div_floor(&x)).shr(1_i32);
+    }
+    if !(x.pow(2) <= *n && *n < (x.clone() + bigint!(1)).pow(2)) {
```

Comment:
I think we don't need any of the `x.clone()`s. Only `n` needs to be cloned due to it being a ref.

In any case, it may be better to call `n.sqrt()`. It's defined for `BigInt`.

Reply:
I could remove the clone on the comparison, but removing the rest of the clones() in x is not possible (at least not without further changes)

Reply:
Why not?

Reply:
The hint specifically uses this version of square root (there is this version and a normal square root in math_utils)

Reply:
I think that answers the other comment.